transpose.hhs

```diff
@@ -1,40 +1,55 @@
 /**
  * @author Jason Reynolds
- * @param input - the matrix to transpose, aka make columns->rows and rows->columns. Must be square, 2d
+ * @modifiedBy Jing Stone
+ * @param input - the matrix to transpose, aka make columns->rows and rows->columns. Support list and 2d matrix.
  * @returns - the transposed matrix, as a Mat object
  * 
  * Simple function that transposes a matrix and returns it. This does NOT modify the input 
  */
 
 
-
 function transpose(input) {
 
     *import math: ndim
     *import math: deep_copy
 
     //declare raw_in to be input clone if Mat and input deep copy otherwise
     let in_type = (input instanceof Mat);
-    let raw_in = (in_type) ? input.clone() : deep_copy(input);
-
     //if input is Mat degrade to JS array to have the below code work for both JS arrays and Mat objects
-    if (input instanceof Mat) {
-        raw_in = raw_in.val;
-    }
+    /**
+     * the speed the mat.clone() is a little slow, need some optimization.
+     */
+    let raw_in = (in_type) ? input.clone().val : deep_copy(input);
+
 
     //checks: 0 dims, not square, ndim not === 2? 
-    if (raw_in.length === 0 || raw_in[0].length === 0 || !(raw_in.length === raw_in[0].length) || !(ndim(raw_in) === 2)) {
-        throw new Error('Matrix must be non-empty, square and 2d');
-    }
+    if (raw_in.length === 0 || raw_in[0].length === 0 || (ndim(raw_in) > 2)) {
```

Comment:
I'd still think you need  the raw_in.length !== raw_in[0].length to make sure it's square, otherwise you can have a matrix like [4,0,0,0],[3,3,3],[2,2,2] and the 4,0,0,0 row will give a column of 4,0,0, undefined (a problem moreso with the mat class but we should also cover it in functions i think)
```suggestion
    if (raw_in.length === 0 || raw_in[0].length === 0 || (ndim(raw_in) > 2) || !(raw_in.length === raw_in[0].length)) {
```
let me know what you think